src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+                    try self.frame.?.renderToPixels(&pixels_opt.*.?);
+                },
+
+                .SOF1 => return error.UnsupportedFrameFormat,
+                .SOF2 => return error.UnsupportedFrameFormat,
+                .SOF3 => return error.UnsupportedFrameFormat,
+                .SOF5 => return error.UnsupportedFrameFormat,
+                .SOF6 => return error.UnsupportedFrameFormat,
+                .SOF7 => return error.UnsupportedFrameFormat,
+                .SOF9 => return error.UnsupportedFrameFormat,
+                .SOF10 => return error.UnsupportedFrameFormat,
+                .SOF11 => return error.UnsupportedFrameFormat,
+                .SOF13 => return error.UnsupportedFrameFormat,
+                .SOF14 => return error.UnsupportedFrameFormat,
+                .SOF15 => return error.UnsupportedFrameFormat,
+
+                .DefineQuantizationTables => {
+                    try self.parseDefineQuantizationTables(reader);
+                },
+
+                .Comment => {
+                    if (JPEG_DEBUG) std.debug.print("Skipping comment segment\n", .{});
+
+                    const comment_length = try reader.readIntBig(u16);
+                    try seek_stream.seekBy(comment_length - 2);
+                },
+
+                else => {
+                    return error.UnknownMarker;
+                },
+            }
+        }
+
+        return if (self.frame) |frame|
+            ImageInfo{
+                .width = frame.frame_header.samples_per_row,
+                .height = frame.frame_header.row_count,
+            }
+        else
+            error.FrameHeaderDoesNotExist;
+    }
+
+    // Format interface
+
+    pub fn formatInterface() FormatInterface {
+        return FormatInterface{
+            .format = @ptrCast(FormatInterface.FormatFn, format),
+            .formatDetect = @ptrCast(FormatInterface.FormatDetectFn, formatDetect),
+            .readForImage = @ptrCast(FormatInterface.ReadForImageFn, readForImage),
+            .writeForImage = @ptrCast(FormatInterface.WriteForImageFn, writeForImage),
+        };
+    }
+
+    fn format() ImageFormat {
+        return ImageFormat.Jpeg;
+    }
+
+    fn formatDetect(reader: ImageReader, seek_stream: ImageSeekStream) !bool {
+        const maybe_start_of_image = try reader.readIntBig(u16);
+        if (maybe_start_of_image != @enumToInt(Markers.StartOfImage)) {
+            return false;
+        }
+
+        try seek_stream.seekTo(6);
+        var identifier_buffer: [4]u8 = undefined;
+        _ = try reader.read(identifier_buffer[0..]);
+
+        return std.mem.eql(u8, identifier_buffer[0..], "JFIF");
+    }
+
+    fn readForImage(allocator: *Allocator, reader: ImageReader, seek_stream: ImageSeekStream, pixels_opt: *?color.ColorStorage) !ImageInfo {
+        var jpeg = JPEG.init(allocator);
+        defer jpeg.deinit();
+
+        return try jpeg.read(reader, seek_stream, pixels_opt);
```

Comment:
only readForImage should return an ImageInfo struct. It is only used for when you are using the Image abstraction. The library is designed around the idea that you use Image when you don't care for the image format, or you can use the format implementation directly for more finer control. Will be useful for some image application that want to access PNG chunks for example.

Reply:
Hm... I suppose I could return the frame directly from `jpeg.read`? That should contain all the necessary pieces for someone to go in and mess around if they please. Then here I just do what I do at the end of `JPEG.read` right now.

Reply:
Done.